### PR TITLE
feat(extensions): global data ingestion

### LINKED ISSUE #14: feat: custom log to kinesis firehose
Have discussed with @titanjer and we are interested to build a custom extension to pipeline custom log to backend data ingestion service like **Amazon Kinesis Firehose**. The high level API may look like

```ts

const stream = new firehose.CfnDeliveryStream(...);

const extension = new CustomFirehoseLog(scope, id, {
    targetStream: stream,
    includeBody: true,
    requiredHeaders,
});
```

### Use Case

https://aws.amazon.com/tw/blogs/networking-and-content-delivery/global-data-ingestion-with-amazon-cloudfront-and-lambdaedge/

圖片

### Other

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by titanjer**
Fixes #14

### CODE REVIEW

File: src/__tests__/integ.global-data-ingestion.test.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      roleArn: firehoseRole.roleArn,
+      errorOutputPrefix: 'error',
+    },
+  });
+
+  /**
+   * Cloudfront and Lambda@Edge
+   */
+  const ext = new extensions.GlobalDataIngestion(stack, 'GlobalDataIngestion', {
+    firehoseStreamName: firehoseStream.deliveryStreamName ?? 'gg',
```
Comment (pahud) [inline src/__tests__/integ.global-data-ingestion.test.ts:50]:
```suggestion
    firehoseStreamName: firehoseStream.deliveryStreamName!
```
I believe the `deliveryStreamName` should exist as you already defined it in line 14 and 33.

File: src/__tests__/integ.global-data-ingestion.test.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  });
+
+  /**
+   * Outputs
+   */
+  new cdk.CfnOutput(stack, 'distributionDomainName', {
+    value: dist.distributionDomainName,
+  });
+  new cdk.CfnOutput(stack, 'firehoseStreamName', {
+    value: firehoseStream.deliveryStreamName ?? 'gg',
```
Comment (pahud) [inline src/__tests__/integ.global-data-ingestion.test.ts:71]:
```suggestion
    value: firehoseStream.deliveryStreamName!,
```

File: src/demo/global-data-ingestion/demo.sh [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  curl $URL --request POST --data "${data}"
+  sleep 1
+done
+
+echo
+BUCKET=$(aws cloudformation describe-stacks \
+        --stack-name global-data-ingestion \
+        --query "Stacks[0].Outputs[?starts_with(OutputKey, 'firehoseDataBucket')].OutputValue" \
+        --output text)
+echo "Please open S3 bucket '${BUCKET}' to see the results"
```
Comment (pahud) [inline src/demo/global-data-ingestion/demo.sh:22]: NICE!

File: src/demo/global-data-ingestion/index.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+});
+
+/**
+ * Outputs
+ */
+new cdk.CfnOutput(stack, 'distributionDomainName', {
+  value: dist.distributionDomainName,
+});
+new cdk.CfnOutput(stack, 'firehoseStreamName', {
+  value: firehoseStream.deliveryStreamName ?? 'gg',
```
Comment (pahud) [inline src/demo/global-data-ingestion/index.ts:70]:
```suggestion
  value: firehoseStream.deliveryStreamName!,
```

File: src/extensions.ts [hunk 5]
```diff
@@ -26,6 +27,10 @@ export interface IExtensions {
    * The Lambda edge event type for this extension
    */
   readonly eventType: cf.LambdaEdgeEventType;
+  /**
+   * Allows a Lambda function to have read access to the body content.
+   */
```
Comment (pahud) [inline src/extensions.ts:32]: Add a `@deafult` tag here as it's optional property.

**Review by pahud — CHANGES_REQUESTED** [review]
Nice! Let's address some minor issues in the comments.

**Comment by pahud** [comment]
HI @guan840912  please help review if you have time. thanks.

**Review by guan840912 — APPROVED** [review]

File: src/__tests__/integ.global-data-ingestion.test.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+  /**
+   * Cloudfront and Lambda@Edge
+   */
+  const ext = new extensions.GlobalDataIngestion(stack, 'GlobalDataIngestion', {
+    firehoseStreamName: firehoseStream.deliveryStreamName ?? 'gg',
+  });
+
+  // const func = lambda.Function.fromFunctionArn(stack, 'LambdaEdgeFunction', ext.functionArn);
+  // func.role?.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonKinesisFirehoseFullAccess'));
```
[the remark this answers is not in the document]
Comment (titanjer) [reply src/__tests__/integ.global-data-ingestion.test.ts:54]: ok, btw this notes this can not works, I dont know why.

**Review by guan840912 — APPROVED** [review]

**Pull request merged**